comit/src/order.rs

```diff
@@ -89,27 +84,93 @@ impl BtcDaiOrder {
             created_at: OffsetDateTime::now_utc(),
         }
     }
+}
+
+/// A newtype representing a quantity in a certain base currency B.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Quantity<B> {
+    value: B,
+}
+
+impl Quantity<Bitcoin> {
+    pub fn new(value: Bitcoin) -> Self {
+        Self { value }
+    }
+
+    pub fn sats(&self) -> u64 {
+        self.value.as_sat()
+    }
+
+    /// The [`Bitcoin`] type encapsulates sats and btc well, hence we can just
+    /// provide access to the inner value here.
+    pub fn to_inner(&self) -> Bitcoin {
+        self.value
+    }
+}
+
+#[cfg(test)]
+pub fn btc(btc: f64) -> Quantity<Bitcoin> {
+    Quantity::new(Bitcoin::from_sat(
+        bitcoin::Amount::from_btc(btc).unwrap().as_sat(),
+    ))
+}
+
+/// A newtype representing the price of one unit of the base currency B in the
+/// quote currency Q.
+///
+/// The core idea around of this type is to enforce the unit of the rate when
+/// doing calculations. We achieve that by adding "loud" constructors and
+/// accessors for combinations of base and quote currency.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Price<B, Q> {
+    value: Q,
+    _base: PhantomData<B>,
+}
 
-    /// Returns the price of this order in the given denomination.
+impl Price<Bitcoin, Erc20Quantity> {
+    /// Constructs a new instance of Price where the rate is given in WEI/SAT.
     ///
-    /// The field `price` holds the value in WEI per SATOSHI, hence we need to
-    /// multiply it by 100_000_000 to get WEI per BTC.
-    pub fn price(&self, denom: Denomination) -> Erc20Quantity {
-        let price = self.price.clone();
-        match denom {
-            Denomination::WeiPerSat => price,
-            Denomination::WeiPerBtc => 100_000_000 * price,
+    /// This is how we store the data internally and hence we don't need to do
+    /// any conversions.
+    pub fn from_wei_per_sat(rate: Erc20Quantity) -> Self {
+        Price {
+            value: rate,
```

Comment:
Can this please be call `price`, or `inner`, or `p` for the same reasons as above - no need to have a second name for the same thing here :)